### PR TITLE
Null-aware GroupKeyIndex

### LINKED ISSUE #1762: IndexScan crashes when columns are pruned
After adding indexes, the TPC-H benchmark crashes when an index scan is used and columns are pruned. Probably the column ids are not updated and [column_ids_after_pruning](https://github.com/mweisgut/hyrise/blob/d3f338e6f947df5239ee1be40dc2f6e974ac65b2/src/lib/utils/column_ids_after_pruning.cpp) might help.

Benchmark flags:
`./build/hyriseBenchmarkTPCH -s 0.001 -r 1`

- [Failing assertion](https://github.com/mweisgut/hyrise/blob/a0415b8d8ed5aeac97e2b7b668bbddebef4ec199/src/lib/operators/index_scan.cpp#L111)

- [How indexes were added](https://github.com/mweisgut/hyrise/blob/d5049f8190ac5efbf2e93e2fdee411a6fb014f10/src/benchmarklib/abstract_table_generator.cpp#L46-L96)

### PR DESCRIPTION
**Pull request by mweisgut**
Fixes #1762

### CODE REVIEW

File: src/lib/storage/index/group_key/group_key_index.cpp [hunk 1]
```diff
@@ -20,9 +20,9 @@ GroupKeyIndex::GroupKeyIndex(const std::vector<std::shared_ptr<const BaseSegment
   Assert((segments_to_index.size() == 1), "GroupKeyIndex only works with a single segment.");
 
   // 1) Initialize the index structures
-  // 1a) Set the index_offset to size of the dictionary + 1 (plus one to mark the ending position)
+  // 1a) Set the index_offset to size of the dictionary + 2 (plus one for null and plus one to mark the ending position)
   //     and set all offsets to 0
-  _index_offsets = std::vector<size_t>(_indexed_segments->unique_values_count() + 1u, 0u);
+  _index_offsets = std::vector<size_t>(_indexed_segments->unique_values_count() + 2u, 0u);
```
Comment (mrks) [inline src/lib/storage/index/group_key/group_key_index.cpp]: Add a comment explaining that and how the rest of NULL value handling happens automagically?

File: src/test/storage/group_key_index_test.cpp [hunk 2]
```diff
@@ -11,15 +11,32 @@
...
 
 namespace opossum {
 
 class GroupKeyIndexTest : public BaseTest {
  protected:
   void SetUp() override {
-    dict_segment = BaseTest::create_dict_segment_by_type<pmr_string>(
-        DataType::String, {"hotel", "delta", "frank", "delta", "apple", "charlie", "charlie", "inbox"});
+    auto value_segment_str = std::make_shared<ValueSegment<pmr_string>>(true);
+    //                                          corresponding ValueID
```
Comment (mweisgut) [inline src/test/storage/group_key_index_test.cpp]: Position in corresponding attribute vector

File: src/lib/storage/index/group_key/group_key_index.cpp (hunk 1 above)
Comment (Bensk1) [reply src/lib/storage/index/group_key/group_key_index.cpp]: This has nothing to do with the current PR but now that I see it: the `_index_offsets` points into the `_index_postings`. The `_index_postings` has the same size as the indexed segment which should be at maximum `ChunkOffset` which is a 4 byte instead an 8 byte (`size_t`) integer, right? Could we save memory here?

Comment (Bensk1) [reply src/lib/storage/index/group_key/group_key_index.cpp]: Maybe you could also add the explanation to the schematic explanation in the header file.

File: src/test/storage/group_key_index_test.cpp [hunk 3]
```diff
@@ -11,15 +11,32 @@
...
+    value_segment_str->append("delta");     //  2
+    value_segment_str->append("frank");     //  3
+    value_segment_str->append("delta");     //  4
+    value_segment_str->append(NULL_VALUE);  //  5
+    value_segment_str->append(NULL_VALUE);  //  6
+    value_segment_str->append("apple");     //  7
+    value_segment_str->append("charlie");   //  8
+    value_segment_str->append("charlie");   //  9
+    value_segment_str->append("inbox");     // 10
+    value_segment_str->append(NULL_VALUE);  // 11
```
Comment (mrks) [inline src/test/storage/group_key_index_test.cpp]: Let this be reflective of the example in the hpp? Also, the example is missing the offset to the end of the postings list. That was broken before.

File: src/test/operators/join_index_test.cpp [hunk 4]
```diff
@@ -130,8 +130,7 @@ class JoinIndexTest : public BaseTest {
       _table_wrapper_m, _table_wrapper_n;
 };
 
-typedef ::testing::Types<AdaptiveRadixTreeIndex, CompositeGroupKeyIndex, BTreeIndex /* , GroupKeyIndex */>
-    DerivedIndices;
+typedef ::testing::Types<AdaptiveRadixTreeIndex, CompositeGroupKeyIndex, BTreeIndex, GroupKeyIndex> DerivedIndices;
```
Comment (mrks) [inline src/test/operators/join_index_test.cpp]: I just realized that the index tests do not cover NULL values. Scrolling through the code, I worry that this PR is breaking more than it is fixing. For example, wouldn't a scan on `WHERE a > 5` also include all NULL values? Same for the join? If this is the case and it still survives the tests, the tests are insufficient.

Maybe we should discuss this in person together with @Bensk1?

File: src/test/storage/group_key_index_test.cpp (hunk 3 above)
Comment (mweisgut) [reply src/test/storage/group_key_index_test.cpp]: https://github.com/hyrise/hyrise/pull/1772/commits/068db8c262ffea70e14801ccdbfb89f00f5085b5

File: src/lib/storage/index/group_key/group_key_index.cpp (hunk 1 above)
Comment (mweisgut) [reply src/lib/storage/index/group_key/group_key_index.cpp]: > Maybe you could also add the explanation to the schematic explanation in the header file.

https://github.com/hyrise/hyrise/commit/068db8c262ffea70e14801ccdbfb89f00f5085b5

> This has nothing to do with the current PR but now that I see it: the `_index_offsets` points into the `_index_postings`. The `_index_postings` has the same size as the indexed segment which should be at maximum `ChunkOffset` which is a 4 byte instead an 8 byte (`size_t`) integer, right? Could we save memory here?

In the worst-case scenario, each value of the segment is a distinct value. Therefore `_index_offsets` has the same size as `_index_postings`. So `ChunkOffset` should be sufficient, shouldn't it?

Comment (mweisgut) [reply src/lib/storage/index/group_key/group_key_index.cpp]: https://github.com/hyrise/hyrise/pull/1772/commits/a05d56d2a1df68c3dddd95fe71b776f5b9cdaf9d

File: src/test/operators/join_index_test.cpp (hunk 4 above)
Comment (mweisgut) [reply src/test/operators/join_index_test.cpp]: Your description sounds similar to this issue https://github.com/hyrise/hyrise/issues/670

I also checked  `index_scan_test` and `index_join_test` and couldn't find any null tests.
The only test which takes null values into account is the referenced one below, but it is disabled and even if we enable it, the used table `resources/test_data/tbl/join_operators/int_outer_join_null.tbl`  is not available anymore. https://github.com/hyrise/hyrise/blob/1a2397cdb9297231efa8b44b5693ee04feae7b80/src/test/operators/join_index_test.cpp#L262-L266

**Comment by mweisgut** [comment]
Moved to #1783

**Pull request closed without merging**